Userland/Libraries/LibWeb/HTML/HTMLCanvasElement.cpp

```diff
@@ -51,6 +54,20 @@ void HTMLCanvasElement::visit_edges(Cell::Visitor& visitor)
         });
 }
 
+void HTMLCanvasElement::apply_presentational_hints(CSS::StyleProperties& style) const
+{
+    // https://html.spec.whatwg.org/multipage/rendering.html#attributes-for-embedded-content-and-images
+    // The width and height attributes map to the aspect-ratio property on canvas elements.
+
+    if (!attribute(HTML::AttributeNames::width).is_null() || !attribute(HTML::AttributeNames::height).is_null())
+        style.set_property(CSS::PropertyID::AspectRatio,
+            CSS::StyleValueList::create(CSS::StyleValueVector {
+                                            CSS::IdentifierStyleValue::create(CSS::ValueID::Auto),
+                                            CSS::RatioStyleValue::create(CSS::Ratio { static_cast<double>(width()), static_cast<double>(height()) }) },
```

Comment:
I don't think using `width()` and `height()` is quite right. Those always return a value, regardless of whether the attribute was good, so we don't know if both attributes contain valid numbers. (The check for null above only makes sure that they have values at all.) eg, `<canvas width="yellow" height="bananas">`.

From [the spec](https://html.spec.whatwg.org/multipage/rendering.html#map-to-the-aspect-ratio-property):

> When the text below says that a pair of attributes w and h on an element element map to the aspect-ratio property, it means that if element has both attributes w and h, and parsing those attributes' values using the [rules for parsing non-negative integers](https://html.spec.whatwg.org/multipage/common-microsyntaxes.html#rules-for-parsing-non-negative-integers) doesn't generate an error for either, then the user agent is expected to use the parsed integers as a [presentational hint](https://html.spec.whatwg.org/multipage/rendering.html#presentational-hints) for the ['aspect-ratio'](https://drafts.csswg.org/css-sizing-4/#aspect-ratio) property of the form auto w / h.

(It might actually be worth making reusable functions/methods for these "map to the foo" things. Like you say, creating this StyleValue by hand is very awkward.)

Reply:
My bad, didn't read into that part of spec. I'll also look into the spec behavior of the height and width attributes converting to integers for the height and width.

Reply:
No worries, there's a _lot_ of spec!

Reply:
All spec-ed out